src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs

```diff
@@ -896,10 +914,25 @@ private ZipArchiveEntry CreateEntry(ZipArchive package, string entryName, Compre
             return entry;
         }
 
-        private ZipArchiveEntry CreatePackageFileEntry(ZipArchive package, string entryName, DateTimeOffset timeOffset, CompressionLevel compressionLevel)
+        private ZipArchiveEntry CreatePackageFileEntry(ZipArchive package, string entryName, DateTimeOffset timeOffset, CompressionLevel compressionLevel, StringBuilder warningMessage)
         {
             var entry = package.CreateEntry(entryName, compressionLevel);
-            entry.LastWriteTime = timeOffset;
+
+            if (timeOffset.UtcDateTime < ZipFormatMinDate)
+            {
+                warningMessage.AppendLine(StringFormatter.ZipFileTimeStampModified(entryName, timeOffset.DateTime.ToShortDateString(), ZipFormatMinDate.ToShortDateString()));
+                entry.LastWriteTime = ZipFormatMinDate;
+            }
+            else if (timeOffset.UtcDateTime > ZipFormatMaxDate)
+            {
+                warningMessage.AppendLine(StringFormatter.ZipFileTimeStampModified(entryName, timeOffset.DateTime.ToShortDateString(), ZipFormatMaxDate.ToShortDateString()));
+                entry.LastWriteTime = ZipFormatMaxDate;
+            }
+            else
+            {
+                entry.LastWriteTime = timeOffset;
```

Comment:
> But for actual physical files it's already in UTC timezone and it comes from below call stack.

Yes, but `NuGet.Packaging` is a package that customers can reference in their own projects, and by implementing their own `IPackageFile` that does not return UTC time, they could encounter problems. You've fixed it by using `.UtcDateTime`, but my point is that since NuGet is not only a tool, but also a package/SDK, we need to consider all the different possible usage.